### DISCUSSION TITLE
LangChain compatibility?

### DISCUSSION THREAD

**Discussion by erenirmak** [MeetKai/functionary#171] [General]
Is it possible to integrate this model with the LangChain environment? Can it use LangChain tools?

**Comment by noshila** [comment]
Hey, any updates?

**Comment by khai-meetkai** [comment]
@noshila I think we can integrate this model with LangChain environment, as we can deploy functionary as **an openai-compatible service**. So any library, program, software that is using OpenAI can switch to using functionary as an alternative

**Comment by unfailingsalvage1448** [comment]
I tried it with vllm, using the langchain class ChatOpenAI, but got:

BadRequestError: Error code: 400 - {'object': 'error', 'message': "FunctionaryTokenizer.apply_chat_template() missing 1 required positional argument: 'tools'", 'type': 'BadRequestError', 'param': None, 'code': 400}

then I tried passing the tools as a model_kwargs "tools", but I get

BadRequestError: Error code: 400 - {'object': 'error', 'message': "[{'type': 'extra_forbidden', 'loc': ('body', 'tools'), 'msg': 'Extra inputs are not permitted', 

Not sure if you can fix this using the vanilla Langchain ChatOpenAI class. So you might need to develop a dedicated langchain class for this model.

Langchain is kind of esoteric when it comes to different model prompt templates, especially tool usage. I don't understand it at all. It would be cool if it worked the same as in the transformers library. But oc that is nothing you can fix.

**Comment by dwight-foster** [comment]
I have added functionary v3.1 small to ollama. You should be able to use the ollama integration in langchain to run the model and execute langchain functions. Here is the model link: https://ollama.com/dwightfoster03/functionary-small-v3.1. I have not tested it with langchain yet but it should work.

**Reply by jeffreymeetkai** [reply]
The template is wrong. We do not use this template in fine-tuning. We use the "User-defined Custom tool calling" format [here](https://llama.meta.com/docs/model-cards-and-prompt-formats/llama3_1/) with some adjustments on top of it.

We have looked at ollama integration but felt that it is not feasible currently as all our other models require converting json schema into Typescript schema. We will need to implement this rather complex conversion in the Go template syntax which may not be possible.

**Reply by dwight-foster** [reply]
I did not realize that thank you. I messed around with the prompt format this one just worked for me but I will add that to the readme. One question doesn’t 3.1 just use the same template as llama 3.1? That is why I chose it over 3.2 but I could be wrong.

**Reply by jeffreymeetkai** [reply]
Yes, it uses one of the templates in llama 3.1 but the documentation link shows that there are 2 prompt formats that can be used, either the tool definitions are in the user prompt or the system prompt (we chose the latter and made some more adjustments to it such as removing the built-in tools and date as llama3.1-instruct has been overfitted to calling those built-in tools).

All other Functionary models except for v3.1 uses our own prompt format which involves converting tool definitions from JSON schema to TS schema. I have managed to create working code for this in Jinja chat template as part of an effort to increase integration across a larger range of inference frameworks but ollama uses Go template syntax which I think is less powerful than Jinja. Thus, we are not sure if it is feasible to integrate all our models except v3.1 into ollama currently.

**Comment by sarahlmk** [comment]
Hi @noshila @erenirmak , sorry for the late update. Now we can use the functionary model in LangChain just like the OpenAI model. You just need to deploy the service and set up the base_url to point to your service. 

Here’s an example of how you might implement it:
```
from langchain_core.tools import tool
from langchain.chat_models import init_chat_model
import getpass
import os

os.environ["OPENAI_API_KEY"] = "functionary"
os.environ["OPENAI_BASE_URL"] = "http://localhost:8000/v1"

model = init_chat_model("meetkai/functionary-v4r-small-preview", model_provider="openai")

@tool
def multiply(a: int, b: int) -> int:
    """Multiply a and b."""
    return a * b

model_with_tools = model.bind_tools([multiply])
result = model_with_tools.invoke("What is 2 multiplied by 3?")

print(result.content)
print(result.tool_calls)
```
